cmd/traffic/cmd/manager/service.go

```diff
@@ -268,10 +272,98 @@ func (s *service) Remain(ctx context.Context, req *rpc.RemainRequest) (*empty.Em
 func (s *service) Depart(ctx context.Context, session *rpc.SessionInfo) (*empty.Empty, error) {
 	ctx = managerutil.WithSessionInfo(ctx, session)
 	sessionID := tunnel.SessionID(session.GetSessionId())
-
+	err := s.UpdateLastEngagementTime(ctx, sessionID)
 	// There's no reason for the caller to wait for this removal to complete.
 	go s.state.RemoveSession(context.WithoutCancel(ctx), sessionID)
-	return &empty.Empty{}, nil
+	return &empty.Empty{}, err
+}
+
+func (s *service) removeUnusedAgent(ctx context.Context, sessionID tunnel.SessionID) error {
+	maxIdleTime := managerutil.GetEnv(ctx).AgentMaxIdleTime
+	if maxIdleTime == 0 {
+		// default aka not set is 0, we don't ever remove agents, skip
+		return nil
+	}
+	agentStateYAML := s.configWatcher.GetAgentStateYaml(ctx)
+	var agentStates map[string]AgentState
+	err := yaml.Unmarshal([]byte(agentStateYAML), &agentStates)
+	if err != nil {
+		return fmt.Errorf("err in unmarshalling YAML: %w", err)
+	}
+	agentKey, err := s.getAgentKey(sessionID)
+	if err != nil {
+		return err
+	}
+
+	lastEngagementTime := agentStates[agentKey].LastEngagementTime
+	idleTime := time.Since(lastEngagementTime)
+	if idleTime > maxIdleTime {
+		s.state.RemoveAgentSession(ctx, sessionID)
+	}
+	return nil
+}
+func (s *service) UpdateLastEngagementTime(ctx context.Context, sessionID tunnel.SessionID) error {
+	namespace := managerutil.GetEnv(ctx).ManagerNamespace
+
+	client := k8sapi.GetK8sInterface(ctx).CoreV1()
+	configMap, err := client.ConfigMaps(namespace).Get(ctx, agentconfig.ManagerAppName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("Error in getting traffic-manager configmap, err: %w", err)
```

Comment:
Instead of "Error in", which starts with a capital letter, use "failure when". See [Errors](https://go.dev/wiki/Errors) in the Go Wiki.